### ISSUE TITLE
Implement some sort of beta testing framework or strategy to vet interface/behavior changes

### ISSUE THREAD

**Issue by blackjackkent** [new-xkit/XKit#746]
In the wake of the chaos that went along with the dropdown-to-buttons change (which is on me for not considering the importance that the "Schedule Post" use case has), I feel we'd probably do well to have a mechanism for vetting major behavior/interface changes in an extension beta before forcing them on all users of the interface.

The idea currently bouncing in my head is to have the beta version of an extension in a separate file (i.e. `editable-reblogs.beta.js`) and a second `gallery.json` (`gallery-beta.json` or whatever) which only displays extensions for which a `.beta.js` file exists.

The user would then have to specifically disable the main extension and install and enable the beta one in order to test it.

This is very rough so any further opinions on the subject would be greatly appreciated.

**Comment by Wolvan** [comment]
I personally suggest of having a switch in the XKit Options, like "Opt in to beta extensions" which just automatically distributes beta extensions to users that have opted in
And maybe instead of another extensions for beta, how about we just have a beta subfolder?

**Comment by bvtsang** [comment]
If we do have a beta channel, the extensions will probably need to be versioned differently to denote its this-is-a-work-in-progress-ness. For example: `3.0.2-beta.1`

**Comment by Wolvan** [comment]
@bvtsang if we already put them somewhere else/rename them, why would you want other denotions? If the beta flag is switched, beta extensions get downloaded

**Comment by hobinjk** [comment]
I believe that implementing this as a separate "XKit Extensions Beta" repo under new-xkit is the best solution. This would require some iffy git tricks but allows us to do a straightforward XKit.download.page override. It would also allow us to practice the type of reorganization described in the 8.0.0 plan where extensions become their own repository.

**Comment by Wolvan** [comment]
@hobinjk We did decide on each extensions becoming their own repository?

**Comment by bvtsang** [comment]
I want to denote the `beta` in the version because:

1. It makes it evident to the user that they're using a beta version. Imagine switching the beta flag and not getting any other indication that you're actually on the beta channel.
2. Our development process won't change significantly. We can still change the versions like normal (e.g. `3.0.2-beta.1` to `3.0.2-beta.2`) and the user will automatically update to the new version.

However, it's possible that I'm misunderstanding how this beta channel process might play out.

**Comment by hobinjk** [comment]
@bvtsang The only drawback I see to that is rewriting the version-parsing code (wherever that lives). Otherwise it is definitely a workable solution

**Comment by Wolvan** [comment]
@hobinjk version parsing's my responsibility, I can definitely work something out here.

@bvtsang But where would that version show?

**Comment by bvtsang** [comment]
The same place where we currently show the extension version.

**Comment by Wolvan** [comment]
We could easily show a BETA message next to the version if the beta switch is flipped in this case

**Comment by bvtsang** [comment]
That could work as well.

I guess I had something different in mind then. Ultimately I think I'm fine with whatever seems reasonable.

**Comment by Wolvan** [comment]
So am I, we just have to figure out the specifics. Version parsing, as I said, I can do with ease. I'd say 2 gallery files would be probably be the easiest/most efficient way

**Comment by hobinjk** [comment]
@Wolvan I agree, special-casing the .beta.js, .beta.css, and .beta.icon.js (for consistency or something) is the best way of implementing this quickly.

**Comment by Wolvan** [comment]
Not sure about the file-endings just yet, though. We could also keep the files in a beta subfolder, which automatically uses the extensions filename and would make auto-updating happen

**Comment by 0xazure** [comment]
> The user would then have to specifically disable the main extension and install and enable the beta one in order to test it.

I'm not a big fan of this, because we already have enough problems with users installing the "right" versions of extensions.  I'd much prefer a switch, as Wolvan suggested.  It would be rather straight-forward to implement this switch switching the download path to point to a `gallery-beta.json`, which could then fall back to non-beta extensions if there is no entry in the `gallery-beta.json` file.  I also think that flipping this switch should automatically re-download any already installed extensions that have a beta version available, an "all-or-nothing" deal rather than on a per-extension basis.

Because I'm not 100% sure on the internal mechanisms, what does installing/uninstalling extensions do to any saved data (settings, lists, etc) an extension may have?  We would probably want the transition between beta and stable channels to be as data-non-destructive as possible.

We would also need to consider how the beta channel interacts with XCloud and data persistence outside of XKit's internal storage.

> practice the type of reorganization described in the 8.0.0 plan where extensions become their own repository

@hobinjk I must have missed this conversation, could you provide a reference?

> denote the `beta` in the version

Yes, this is a must.  The less guessing there is about version numbers and where extension code originated from, the better.

> not getting any other indication that you're actually on the beta channel

We _need_ to make this super-obvious.  I'm thinking a [caution tape](https://www.google.ca/webhp?hl=en#safe=off&hl=en&q=caution+tape) border along the top of the XKit Control Panel, and some obvious indications on the individual beta extensions.

As far as where the files are going to live, this bears some discussion around workflow like #719.  I dislike the idea of maintaining separate files.  Patching in changes from `master` will be a pain, it won't be immediately obvious _when_ there are changes from `master` that need to be integrated, and it feels like we would be exacerbating the issues already identified by #364.  Having a common `beta` branch may be a better option, especially since we would then be able to cherry-pick commits between `master` and `beta` as necessary to keep things up to date.

We also need to consider messaging to our users.  Given the pace at which we tend to push out fixes (generally as soon as possible), how do we communicate that the beta channel is for serious testers only?  How do we prevent the beta channel from just becoming a "flip this switch for the latest changes!" exercise?  Do we want to actively support testers the same way we provide support for our current users, or try and limit it to "tech-savvy" users only?

Lastly, how does a potential beta channel influence the `BETA` flag on extensions, if at all?  Do we want to integrate or coopt this flag in any way?  Does it still have meaning for extensions in the stable channel?

**Comment by Wolvan** [comment]
I absolutely agree with an all or nothing Beta switch, fallback might be a bit hard though. Would we just first poll ``gallery-beta.json`` and then ``gallery.json``, adding anything that is not in the beta file to the gallery as well?

Regarding Uninstall/Install/Update, for now uninstalling does not remove the settings (see #762). I don't think an ``XKit.installed.update()`` call does that either.

Not sure about XCloud, we would need @ChuckL here to answer that.

I really like the caution tape idea :+1: on that!
For something completely obvious in the extension's Panel, how about a ``β`` symbol overlayed over the icons and the text ``BETA`` as some kind of watermark in the extension's seetings page?

@0xazure I have a question about the beta branch: How will we bring that into gh-pages easily? I mean, a seperate beta branch does sound nice but the process of bringing both beta AND master into gh-pages sounds complicated.

Messaging our users is another question I have. Should we add a BETA channel (maybe make it automatically activate with the BETA setting for XKit) in paperboy?
And I'd honestly say we limit beta to the more tech-savy people and above the Beta switch we have provide a huge disclaimer that this is an UNSTABLE branch and thus can and will have tons of bugs.
We could also make another gitter channel (no IRC Bridging here) that is ONLY for beta extensions support. The no IRC bridge means people need a github account, that might stop the less tech savy people from using it

I've never been a real fan of the ``BETA`` extension flag, personally. Does it even do anything in XKit?

Lastly, will our beta-version format be ``MAJOR.MINOR.PATCH-beta1/2/3/4.....``? If so, should I add beta parsing to ``XKit.tools.parse_version()`` to return ``{major, minor, patch, beta}``?

**Comment by 0xazure** [comment]
> Would we just first poll gallery-beta.json and then gallery.json, adding anything that is not in the beta file to the gallery as well?

This is really an implementation detail that we would have to discuss during actual implementation.

> For something completely obvious in the extension's Panel

There's already an icon for "experimental" (see XInbox's 'enable inbox search' option), we could probably re-use that.

> a question about the beta branch...

There's no such thing as "easy" once you start splitting releases into different channels, and our lack of process doesn't really help.  Completely untested idea: we could track the changes in a branch in git, but then leverage the builders to generate actual `.beta.js` files out of the files on the `beta` branch and commit those generated files into `gh-pages`, maybe.  It definitely bears consideration, but we would have to test out some different approaches and ensure we follow proper process.

> add a BETA channel [...] in paperboy

I like this idea.  Giving us a quick and easy way to disseminate information to the testers would be a great asset.  The change wouldn't need to be a part of #554 though, this is something we could add once we've figured out how we're pushing updates.

> beta-version format

SemVer formatting should be in accordance with SemVer 2.0.0.  See [semver#9](http://semver.org/) for semantics for pre-release versions.

The more I consider this, the more I feel it's not going to be an easy thing to implement.  It might have to be a post-8.0.0 project; there are a number of existing issues I see us running into because of existing legacy cruft and tech debt.
